### PR TITLE
feat: add a deep thinking reasoner model (o1-preview/mini)

### PR DESCRIPTION
**Pull request by michaelneale**
This starts to introduce the concept of a "slower" reasoner model (alongside accelerator and processor) - and uses it via a toolkit for enhanced debugging/understanding and code authoring.

It will use the o1 models in some cases as needed, to compliment gpt4-o and gpt-4o-mini in the openai case. The "reasoner" model doesn't do tool calling or planning directly, but is consulting.

### CODE REVIEW

File: src/goose/toolkit/reasoner.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        if isinstance(content, Text):
+            return content
+        else:
+            return Text(str(content))
+
+
+    @tool
+    def deep_debug(self, problem:str) -> str:
+        """
+        Thius tool can assist with debugging when there are errors or problems when trying things
```
Comment (wesrblock) [inline src/goose/toolkit/reasoner.py]: thius -> this

File: tests/curves/p256_tests.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
+// tests/curves/p256_tests.rs
```
Comment (codefromthecrypt) [inline tests/curves/p256_tests.rs]: sorry if obvious, but what's the relationship between this and the reasoner?

Comment (michaelneale) [reply tests/curves/p256_tests.rs]: ah good catch - I was running goose from this dir and having it write some rust code to test out o1.

**Review by wesrblock — APPROVED** [review]
Looks great!

**Comment by michaelneale** [comment]
I have been using this all week to great success.

**Comment by codefromthecrypt** [comment]
@michaelneale curious what specific sort of thing do you feel was more effective or possible with this vs without?

**Comment by michaelneale** [comment]
@codefromthecrypt mostly harder problems - for example the "interactive" fix for goose came from it initially (goose wasn't able to solve things that deep before). It also seems to help avoid goose prematurely jumping to a solution by misunderstanding the nuances of a problem and doing a "too obvious" thing. Well worth it IMO (and as a toolkit - so it doesn't kick in all the time).

**Comment by michaelneale** [comment]
@baxen I think this is ready for review - if interested to include it as something people can opt in or not (as toolkit, no extra deps)

**Review by salman1993 — APPROVED** [review]

**Pull request merged**